File: scripts/ci-proof/route.go
package main

import (
	"fmt"
	"strings"
)

const (
	proofRouteFeatureDev          = "feature_dev"
	proofRoutePromotionMain       = "promotion_main"
	proofRouteReconciliationMain  = "reconciliation_main"
	proofRouteFoundationPromotion = "foundation_promotion"
	proofRouteProtectedPushDev    = "protected_push_dev"
	proofRouteProtectedPushMain   = "protected_push_main"
)

func classifyProofRoute(event, baseRef string) (string, error) {
	switch event + ":" + baseRef {
	case "pull_request:dev":
		return proofRouteFeatureDev, nil
	case "pull_request:main":
		return proofRoutePromotionMain, nil
	case "push:dev":
		return proofRouteProtectedPushDev, nil
	case "push:main":
		return proofRouteProtectedPushMain, nil
	default:
		return "", fmt.Errorf("unsupported CI proof route tuple")
	}
}

func validContextProofRoute(context contextInput) bool {
	if context.Route == proofRouteFoundationPromotion {
		return isFoundationPromotionContext(context) && context.FoundationPromotion != nil && context.FoundationPromotion.HeadSHA == context.HeadSHA
	}
	route, err := classifyProofRouteWithHead(context.Event, context.BaseRef, context.HeadRef)
	return err == nil && context.Route == route && context.FoundationPromotion == nil
}

func validBundleProofRoute(bundle proofBundle) bool {
	if bundle.FoundationPromotion != nil {
		return isFoundationPromotionBundle(bundle)
	}
	route, err := classifyProofRouteWithHead(bundle.Event, bundle.BaseRef, bundle.HeadRef)
	return err == nil && route == expectedProofRoute(bundle.Event, bundle.BaseRef, bundle.HeadRef)
}

func isPromotionContext(context contextInput) bool {
	if context.Route == proofRouteFoundationPromotion || context.FoundationPromotion != nil {
		return false
	}
	route, err := classifyProofRouteWithHead(context.Event, context.BaseRef, context.HeadRef)
	return err == nil && (route == proofRoutePromotionMain || route == proofRouteReconciliationMain)
}

func isPromotionBundle(bundle proofBundle) bool {
	if bundle.FoundationPromotion != nil {
		return false
	}
	route, err := classifyProofRouteWithHead(bundle.Event, bundle.BaseRef, bundle.HeadRef)
	return err == nil && (route == proofRoutePromotionMain || route == proofRouteReconciliationMain)
}

func classifyProofRouteWithHead(event, baseRef, headRef string) (string, error) {
	const prefix = "agent/main-history-reconciliation-"
	if event == "pull_request" && baseRef == "main" && strings.HasPrefix(headRef, prefix) && len(headRef) > len(prefix) {
		return proofRouteReconciliationMain, nil
	}
	if event == "pull_request" && baseRef == "main" && headRef != "" && headRef != "dev" {
		return "", fmt.Errorf("ordinary agent-to-main route is not authorized")
	}
	return classifyProofRoute(event, baseRef)
}

func expectedProofRoute(event, baseRef, headRef string) string {
	route, err := classifyProofRouteWithHead(event, baseRef, headRef)
	if err != nil {
		return ""
	}
	return route
}

func isReconciliationContext(context contextInput) bool {
	return context.Route == proofRouteReconciliationMain && context.Event == "pull_request" && context.BaseRef == "main" && strings.HasPrefix(context.HeadRef, "agent/main-history-reconciliation-")
}

func isReconciliationBundle(bundle proofBundle) bool {
	return bundle.Event == "pull_request" && bundle.BaseRef == "main" && strings.HasPrefix(bundle.HeadRef, "agent/main-history-reconciliation-")
}

File: scripts/ci-proof/route.js
'use strict';

const crypto = require('node:crypto');

const RECONCILIATION_ROUTE = 'reconciliation_main';
const RECONCILIATION_BRANCH_PREFIX = 'agent/main-history-reconciliation-';

const schema = 'gooo/ci-proof-route/v1';
const foundationPromotion = Object.freeze({
  repository: 'kimjooyoon/meta-ontology-go',
  pullRequest: 602,
  headRef: 'agent/foundation-discovery-recovery-20260830',
  baseRef: 'main',
  baseSha: 'cd9727af80f5118405290d3be96890c18e1529c0',
});
const routes = Object.freeze({
  'pull_request:dev': 'feature_dev',
  'pull_request:main': 'promotion_main',
  'push:dev': 'protected_push_dev',
  'push:main': 'protected_push_main',
});

function isFoundationPromotion(input) {
  return input.event === 'pull_request' && input.repository === foundationPromotion.repository &&
    input.prNumber === foundationPromotion.pullRequest && input.headRef === foundationPromotion.headRef &&
    input.baseRef === foundationPromotion.baseRef && input.baseSha === foundationPromotion.baseSha;
}

function isReconciliationPromotion(input) {
  return input.event === 'pull_request' && input.baseRef === 'main' && typeof input.headRef === 'string' && input.headRef.startsWith(RECONCILIATION_BRANCH_PREFIX) && input.headRef.length > RECONCILIATION_BRANCH_PREFIX.length;
}

function classifyProofRoute(event, baseRef, input = {}) {
  if (isFoundationPromotion({...input, event, baseRef})) return 'foundation_promotion';
  if (isReconciliationPromotion({...input, event, baseRef})) return RECONCILIATION_ROUTE;
  if (event === 'pull_request' && baseRef === 'main' && typeof input.headRef === 'string' && input.headRef.startsWith('agent/') && input.headRef !== 'dev' && !input.headRef.startsWith('agent/foundation-discovery-recovery-')) {
    throw new Error('ordinary agent-to-main route is not authorized');
  }
  const route = routes[event + ':' + baseRef];
  if (!route) throw new Error('unsupported CI proof route tuple');
  return route;
}

function buildProofRouteEvidence(input) {
  const payload = {
    schema,
    event: input.event,
    event_ref: input.eventRef,
    base_ref: input.baseRef,
    head_sha: input.headSha,
    route: classifyProofRoute(input.event, input.baseRef, input),
    guardian_required: input.event === 'pull_request' && input.baseRef === 'main' && !isFoundationPromotion(input),
  };
  const digest = crypto.createHash('sha256').update(JSON.stringify(payload)).digest('hex');
  return {...payload, digest: 'sha256:' + digest};
}

function validateProofRouteEvidence(evidence, input) {
  const expected = buildProofRouteEvidence(input);
  if (JSON.stringify(evidence) !== JSON.stringify(expected)) {
    throw new Error('proof route evidence is stale, malformed, or unbound');
  }
}

module.exports = {
  buildProofRouteEvidence,
  classifyProofRoute,
  foundationPromotion,
  isFoundationPromotion,
  isReconciliationPromotion,
  RECONCILIATION_BRANCH_PREFIX,
  RECONCILIATION_ROUTE,
  validateProofRouteEvidence,
};
